### ISSUE TITLE
Bitmapped fonts not working.

### ISSUE THREAD

**Issue by roflcopter4** [bryphe/oni#478]
I'm of the opinion that when it comes to working with any text whatsoever on any background that isn't white, anti aliasing is, as a rule, evil. Especially when the fonts are coloured. These are both conditions well met by any halfway decent colour scheme in Vim, and as a result I greatly prefer to use bitmapped fonts. Not only are they clearer and smaller, but you get no blurring or artifacts due to colouration.

Naturally these fonts generally work in any console vim/neovim, where bitmapped fonts are obviously supported. They also work in the standard gvim and in the QT gui for Neovim. However, they seem not to be supported by Oni at all. The program has completely ignored every bitmapped font I've tried to get it to use. I've tried several, and each at many point sizes (under the assumption that it may be getting confused when the point size doesn't match the one and only point size the font supports) with no luck. I'm not sure whether there is something I may be doing wrong, or if this is a bug, or whether there is simply no desire to support bitmapped fonts at all.

I'm not entirely sure if there is much I can offer in the way of debugging information. If bitmapped fonts work for everyone but me then perhaps I ought to try a little harder there, but if they do not then this should be a fairly simple bug to reproduce. It may be worth noting that I am running it in Windows at the moment (primarily because I'm mostly satisfied with terminal Neovim in linux, whereas there is no such a thing at all in Windows).

**Comment by extr0py** [comment]
Hi @roflcopter4 , thanks for the detailed issue!

I did some research - it looks like bitmap fonts aren't well supported in the browser context. There was a similiar question about using a bitmap font in the browser context: https://stackoverflow.com/questions/5354935/how-to-embed-a-custom-bitmap-font-into-website-using-css

But I don't think that would solve your issue with anti-aliasing, as the solution there was to convert it to a web font (.woff) and use that.

There are a couple of possibilities:
 - __Possibility 1__ - Chromium (and then Electron) supports bitmap fonts, which would mean Oni gets the support for free. I don't see any plans for this, though,  currently.  In fact someone asked a similiar question about Atom here: https://discuss.atom.io/t/will-bitmap-fonts-eventually-be-available/10567

- __Possibility 2__ - Oni supports it independently - this could be done by reviving a canvas-renderer strategy for the editor surface (which was actually the original render strategy). The canvas lends itself well to bitmap font rendering and would make this relatively easy to implement. The canvas renderer had poorer support for things like ligatures, though, and we'd have to document that.

Option 2 is doable, it just depends on the level of interest. If anyone else would find bitmap support useful, please "thumbs up" this issue so we can track.

Also, do you have a recommendation for a bitmap font to test with for this issue?

Thanks!

**Comment by roflcopter4** [comment]
I do have a recommendation for a font. My favorite font in the world at the moment is one called Dina, which can be obtained [here](https://www.donationcoder.com/Software/Jibz/Dina/) for Windows. There is no "official" version compatible with Linux, but it is fairly easy to convert, and someone made a very good conversion for download in [this thread](http://www.donationcoder.com/forum/?topic=7857.0), which I have tested and works flawlessly. 

I love this font because it is compact, easy to read, the characters are all distinct (esp O and 0), and unlike most bitmapped fonts it comes with very good bold, italic, and bold-italic versions. Most bitmapped fonts have to fallback on either terrible bold conversions that blur letters, or automatically generated ones that are even worse, and the same for italic. Dina does not. I even unnecessarily took some [comparison screenshots here](http://imgur.com/a/G8nTK) (you have to open the images full screen to see the difference clearly). The anti-aliasing issue is perhaps not as dramatic as I make it seem, but to me the blurriness is very clearly evident in the non-bitmapped ones and bothers me quite a lot.

Anyway, that's all not particularly relevant. I agree that the solution from stackoverflow is not really applicable here, and also agree that it's fairly unlikely that Chromium would up and decide to support bitmapped fonts. I wish I could offer support for your second suggestion, but I don't feel that my 200 level computer science + "read a few programming books and write scripts for video games" level of skill is likely up to the task here.

**Comment by roflcopter4** [comment]
I just wanted to post here, for the record, that there already seems to be a miraculous and bizarre partial solution to this problem. A series of unfortunate circumstances compelled me to use Microsoft's VS, which I quickly found also did not support raster fonts. I nonetheless vainly googled around for a while to see if there were a way around, and to some shock I found that there apparently is. [This guy wrote an article](http://www.electronicdissonance.com/2010/01/raster-fonts-in-visual-studio-2010.html) on it several years ago, in which he basically discovered that for whatever reason Microsoft decided to support raster fonts in certain East Asian language fonts, while dropping support for the rest. Why they did this is quite unfathomable, but it does allow a rather monstrous hack. Through a long and specific process, a font can be modified to appear to be an East Asian Truetype font with embedded bitmaps which allows it to be displayed.

I tried the font he included in the article, and it didn't work. However, after some more googling, I found that someone had already converted my [aforementioned favorite font, Dina, to this format.](http://chrisrickard.blogspot.ca/2010/03/dina-font-for-visual-studio-2010.html) It did work in Visual Studio, so long as the correct point size is chosen. Despite there being no connection I know of between that editor and this project, I decided to try the font in Oni too. To my great astonishment, it actually works. I have no idea why. Again, so long as the correct size is chosen, the font simply works (Dina supports 11px (8 point), 12px (9 point), and 13px (10 point)).

I don't know whether or not it was worth pointing this out, since it barely qualifies as a solution and comes with an overly long and irrelevant story, but I found the fact that it's even possible quite interesting. How can it be properly displaying the fonts when it apparently really should just not be able to? Can that fact be exploited?

**Comment by extr0py** [comment]
Interesting, thanks for the update!

> Through a long and specific process, a font can be modified to appear to be an East Asian Truetype font with embedded bitmaps which allows it to be displayed.

That's pretty bizarre... I did a skim through the article and it seems like there were some _very_ specific requirements for this to work.

> To my great astonishment, it actually works.

Glad it actually worked, and that someone had already converted your font!

One thing to note is that in the latest release (0.2.7), I've also switched Oni back to the canvas render strategy by default, for performance reasons. But it's unlikely that would've changed here, since supporting the bitmap font is actually up to Chrome's renderer.

Unfortunately I don't know too much about Chromium's rendering internals, but perhaps there is some infrastructure in common with WPF in terms of resolving or rendering the fonts. It is interesting though that both Electron & Visual Studio exhibit the same behavior here. Glad you thought to try it in Oni, too!

In any case I'm glad you found a workaround, and thanks for taking the time to reply with your investigation and the working font - I'm sure that will be useful for others!
